test/unit/mc/test_cholesky.cpp

```diff
@@ -0,0 +1,152 @@
+
+TYPED_TEST_CASE(CholeskyTest, MatrixElementTypes);
+
+std::vector<GlobalElementSize> square_sizes({{10, 10}, {25, 25}, {12, 12}, {0, 0}});
+std::vector<GlobalElementSize> rectangular_sizes({{10, 20}, {50, 20}, {0, 10}, {20, 0}});
+std::vector<TileElementSize> square_block_sizes({{5, 5}, {20, 20}});
+std::vector<TileElementSize> rectangular_block_sizes({{10, 30}, {20, 10}});
+
+TYPED_TEST(CholeskyTest, Correctness) {
+  using Type = TypeParam;
+
+  // Note: The tile elements are chosen such that:
+  // - res_ij = 1 / 2^(|i-j|) * exp(I*(-i+j)),
+  // where I = 0 for real types or I is the complex unit for complex types.
+  // Therefore the result should be:
+  // a_ij = Sum_k(res_ik * ConjTrans(res)_kj) =
+  //      = Sum_k(1 / 2^(|i-k| + |j-k|) * exp(I*(-i+j))),
+  // where k = 0 .. min(i,j)
+  // Therefore,
+  // a_ij = (4^(min(i,j)+1) - 1) / (3 * 2^(i+j)) * exp(I*(-i+j)
+  auto el = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    if (i < j)
+      return TypeUtilities<TypeParam>::element(-9.9, 0.0);
+
+    return TypeUtilities<TypeParam>::polar(std::exp2(-(i + j)) / 3 *
+                                               (std::exp2(2 * (std::min(i, j) + 1)) - 1),
+                                           -i + j);
+  };
+
+  // Analytical results
+  auto res = [](const TileElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    if (i < j)
+      return TypeUtilities<TypeParam>::element(-9.9, 0.0);
+
+    return TypeUtilities<TypeParam>::polar(std::exp2(-std::abs(i - j)), -i + j);
+  };
+
+  for (const auto& size : square_sizes) {
+    for (const auto& block_size : square_block_sizes) {
+      // Matrix to undergo Cholesky decomposition
+      Matrix<Type, Device::CPU> mat(size, block_size);
+      set(mat, el);
+
+      // Matrix with analytical results
+      Matrix<Type, Device::CPU> matres(size, block_size);
+      set(matres, res);
+
+      EXPECT_NO_THROW(cholesky_local(mat, blas::Uplo::Lower, blas::Side::Right, blas::Op::ConjTrans,
+                                     blas::Diag::NonUnit));
+
+      CHECK_MATRIX_NEAR(res, mat, 4 * (mat.size().rows() + 1) * TypeUtilities<TypeParam>::error,
+                        4 * (mat.size().rows() + 1) * TypeUtilities<TypeParam>::error);
+    }
+  }
+}
+
+TYPED_TEST(CholeskyTest, NoSquareMatrixException) {
+  using Type = TypeParam;
+
+  auto el = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    return TypeUtilities<Type>::element(i + 0.001 * j, j - 0.01 * i);
+  };
+
+  // Check for square sizes
+  for (const auto& size : square_sizes) {
+    for (const auto& block_size : square_block_sizes) {
+      Matrix<Type, Device::CPU> mat(size, block_size);
+
+      set(mat, el);
+
+      EXPECT_NO_THROW(cholesky_local(mat, blas::Uplo::Lower, blas::Side::Right, blas::Op::ConjTrans,
+                                     blas::Diag::NonUnit));
+    }
+  }
```

Comment:
This part is not needed. It is already tested by `Correctness` test.

Reply:
Should I check square sizes in `Correctness` or is it better to check this in `NoSquareMatrixException`?

Reply:
The check that the function works for square matrices is implicit in the Correctness check.
(The correctness test checks if the function produces the correct result without throwing).

Moreover this test is likely to fail since the matrix used is not positive definite.